File: day15_part2_2016.rb

def find_time(discs)
  time = 0
  loop do
    break if discs.each_with_index.all? { |(positions, start), i| (start + time + i + 1) % positions == 0 }
    time += 1
  end
  time
end

discs = File.read('input.txt').scan(/Disc #\d+ has (\d+) positions; at time=0, it is at position (\d+)./).map { |positions, start| [positions.to_i, start.to_i] }
puts find_time(discs)
discs << [11, 0]
puts find_time(discs)

File: haskell/day15_part2_2016.hs
import System.IO

data Disc = Disc { positions :: Int, startPos :: Int }

main = do
    contents <- readFile "input.txt"
    let discs = parseInput contents
        newDisc = Disc 11 0
        allDiscs = discs ++ [newDisc]
        time = findFirstTime allDiscs
    print time

parseInput :: String -> [Disc]
parseInput input = map parseLine (lines input)
  where
    parseLine line = let parts = words line
                         positions = read (parts !! 3)
                         startPos = read (init (parts !! 11))
                     in Disc positions startPos

findFirstTime :: [Disc] -> Int
findFirstTime discs = head [t | t <- [0..], all (capsuleFallsThrough t) (zip discs [1..])]

capsuleFallsThrough :: Int -> (Disc, Int) -> Bool
capsuleFallsThrough time (disc, index) = (startPos disc + time + index) `mod` positions disc == 0